packages/relay/src/routes/vote.ts

```diff
@@ -0,0 +1,249 @@
+        },
+    }
+    switch (voteAction) {
+        case VoteAction.UPVOTE:
+            voteCreateStatement.upVote = true
+            postStatement.update.upCount += 1
+            break
+        case VoteAction.DOWNVOTE:
+            voteCreateStatement.downVote = true
+            postStatement.update.downCount += 1
+            break
+        case VoteAction.CANCEL_UPVOTE:
+            postStatement.update.upCount -= 1
+            createVote = false
+            break
+        case VoteAction.CANCEL_DOWNVOTE:
+            postStatement.update.downCount -= 1
+            createVote = false
+            break
+        default:
+            break
+    }
+
+    await addActionCount(db, epochKey, epoch, (txDB) => {
+        let actionCount
+        if (createVote) {
+            txDB.create('Vote', voteCreateStatement)
+            actionCount = 1
+        } else {
+            txDB.delete('Vote', voteDeleteStatement)
+            // epk cancel the action, decrease 1 count
+            actionCount = -1
+        }
+        txDB.update('Post', postStatement)
+
+        return actionCount
+    })
+}
+
+// TODO: should check if user voted to the same post before with other epochKey
+/**
+ * The user can vote or cancel the vote to the post even in the different epoch,
+ * but the user is only allowed to vote either upvote or downvote, not both.
+ *
+ * @param req express request, the body should contain postId, voteAction(refer to types/VoteAction.ts) and zkp
+ * @param res express response
+ * @param db SQLiteConnector
+ * @param synchronizer UnirepSocialSynchronizer
+ * @returns
+ */
+async function Vote(req, res, db: DB, synchronizer: UnirepSocialSynchronizer) {
+    try {
+        //vote for post with _id
+        const { _id, voteAction, publicSignals, proof } = req.body
+
+        // user is able to restore the epochKey from the vote data
+        const epochKeyProof = new EpochKeyProof(
+            publicSignals,
+            proof,
+            synchronizer.prover
+        )
+
+        // get current epoch and unirep contract
+        const epoch = await synchronizer.loadCurrentEpoch()
+
+        validateEpoch(epochKeyProof.epoch, epoch)
+        validateAttesterId(epochKeyProof.attesterId, synchronizer.attesterId)
+        await verifyEpochKeyProof(epochKeyProof)
+
+        // find post which is voted
+        const findPost = await db.findOne('Post', {
+            where: {
+                _id: _id,
+            },
+        })
+        if (!findPost) {
+            throw InvalidPostIdError
+        }
+
+        const epochKey = epochKeyProof.epochKey.toString()
+
+        const findVote = await db.findOne('Vote', {
+            where: {
+                postId: _id,
+                epochKey: epochKey,
+            },
+        })
+
+        verifyVoteAction(voteAction, findVote)
+
+        await exeuteTxs(db, epochKey, epoch, findPost, voteAction)
+
+        res.status(201).json({})
+    } catch (error: any) {
```

Comment:
@jason-ntu This error should have specific type?

Reply:
No it doesn't, since it might be either Error or VoteError.